roachtest: kv/contention/nodes=4 failed

SHA: https://github.com/cockroachdb/cockroach/commits/175c5ada040fd0cbbf178636b1c551d5c2229ec4

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=kv/contention/nodes=4 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1417597&tab=buildLog

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20190802-1417597/kv/contention/nodes=4/run_1
	cluster.go:1726,kv.go:197,cluster.go:2069,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1564726582-33-n5cpu4:5 -- ./workload run kv --init --secondary-index --duration=1h0m0s --cycle-length=512 --concurrency=128 --batch=4 --splits=4 {pgurl:1-4} returned:
		stderr:
		
		stdout:
		    47s        0          553.3          515.9     11.0   1040.2   2281.7  15569.3 write
		     48s        0          203.0          509.4     10.5    637.5   2550.1   4831.8 write
		     49s        0          319.9          505.5      8.4   1409.3   4295.0  24696.1 write
		     50s        0           64.0          496.7      7.6   1946.2   2281.7  16643.0 write
		     51s        0          131.0          489.5     10.5   2080.4   3221.2  15569.3 write
		     52s        0          265.0          485.2      7.1   2147.5   7247.8  13421.8 write
		     53s        0          127.0          478.4      6.6   3355.4   3355.4   8053.1 write
		     54s        0          121.0          471.8     12.1   3489.7  20401.1  24696.1 write
		     55s        0          787.7          477.6     15.7   2013.3   5368.7  21474.8 write
		     56s        0          363.9          475.5     11.0    604.0   3087.0  36507.2 write
		Error: ERROR: result is ambiguous (error=unable to dial n4: breaker open [exhausted]) (SQLSTATE 40003)
		Error:  exit status 1
		: exit status 1
	cluster.go:2090,kv.go:207,test_runner.go:691: unexpected node event: 4: dead

```

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xd9d4e2]

goroutine 173295 [running]:
panic(0x38b2420, 0x6798bc0)
	/usr/local/go/src/runtime/panic.go:565 +0x2c5 fp=0xc003e53d18 sp=0xc003e53c88 pc=0x787af5
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:82
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:390 +0x411 fp=0xc003e53d48 sp=0xc003e53d18 pc=0x79d651
github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace.(*traceSet).Remove(0x0, 0xc0074e5380)
	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:478 +0x22 fp=0xc003e53d70 sp=0xc003e53d48 pc=0xd9d4e2
github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace.(*trace).Finish(0xc0074e5380)
	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:412 +0x137 fp=0xc003e53e40 sp=0xc003e53d70 pc=0xd9d097
github.com/cockroachdb/cockroach/pkg/util/tracing.(*span).FinishWithOptions(0xc003869ce0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/tracer_span.go:316 +0x131 fp=0xc003e53ea8 sp=0xc003e53e40 pc=0xecdf91
github.com/cockroachdb/cockroach/pkg/util/tracing.(*span).Finish(0xc003869ce0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/tracer_span.go:300 +0x67 fp=0xc003e53f08 sp=0xc003e53ea8 pc=0xecde47
github.com/cockroachdb/cockroach/pkg/util/tracing.FinishSpan(0x45de8e0, 0xc003869ce0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/tracer.go:625 +0x40 fp=0xc003e53f20 sp=0xc003e53f08 pc=0xecc410
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*samplerProcessor).Run(0xc000674a80, 0x458cfe0, 0xc007fa6440)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sampler.go:186 +0x13e fp=0xc003e53fa0 sp=0xc003e53f20 pc=0x25ca07e
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).startInternal.func1(0xc00b40ed20, 0xc00387ac60, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:563 +0x68 fp=0xc003e53fc8 sp=0xc003e53fa0 pc=0x25f5708
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1337 +0x1 fp=0xc003e53fd0 sp=0xc003e53fc8 pc=0x7b7891
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).startInternal
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:562 +0x35e
```

We see a crash due to a use-after-finish in a `samplerProcessor`. We saw this twice yesterday in #39239 and in #39190. We had hoped it was fixed by #39203, but it doesn't look like it. Luckily, it looks easy to reproduce.

--- a/pkg/sql/distsqlrun/sample_aggregator.go
+++ b/pkg/sql/distsqlrun/sample_aggregator.go
@@ -135,7 +135,6 @@ func (s *sampleAggregator) pushTrailingMeta(ctx context.Context) {
 func (s *sampleAggregator) Run(ctx context.Context) {
 	s.input.Start(ctx)
 	s.StartInternal(ctx, sampleAggregatorProcName)
-	defer tracing.FinishSpan(s.span)
 
 	earlyExit, err := s.mainLoop(s.Ctx)
 	if err != nil {

--- a/pkg/sql/distsqlrun/sampler.go
+++ b/pkg/sql/distsqlrun/sampler.go
@@ -26,7 +26,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/mon"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
-	"github.com/cockroachdb/cockroach/pkg/util/tracing"
 	"github.com/cockroachdb/errors"
 )
 
@@ -172,7 +171,6 @@ func (s *samplerProcessor) pushTrailingMeta(ctx context.Context) {
 func (s *samplerProcessor) Run(ctx context.Context) {
 	s.input.Start(ctx)
 	s.StartInternal(ctx, samplerProcName)
-	defer tracing.FinishSpan(s.span)
 
 	earlyExit, err := s.mainLoop(s.Ctx)
 	if err != nil {

